src/symbol/sym.rs

```diff
@@ -0,0 +1,104 @@
+//! Implementation of a symbol table entry that will automatically
+//! demangle rustc names.
+
+use addr2line::demangle_auto;
+use object::{SectionIndex, SymbolFlags, SymbolKind, SymbolScope, SymbolSection};
+use std::borrow::Cow;
+
+/// A symbol table entry.
+#[derive(Clone, Debug)]
+pub struct Symbol<'data> {
+    symbol: object::Symbol<'data>,
+}
+
+impl<'data> Symbol<'data> {
+    /// Returns the demangled name if this symbol has a name.
+    #[inline]
+    pub fn name(&self) -> Option<String> {
+        // TODO: Avoid this allocation in every call. (lifetime errors)
+        self.orig_name()
+            .map(|name| demangle_auto(Cow::Borrowed(name), None).to_string())
+    }
+
+    /// Returns the unmangled name of this symbol.
+    #[inline]
+    pub fn orig_name(&self) -> Option<&'data str> {
+        self.symbol.name()
+    }
+
+    /// Return the kind of this symbol.
+    #[inline]
+    pub fn kind(&self) -> SymbolKind {
+        self.symbol.kind()
```

Comment:
I'd suggest adding `impl Deref<Target = object::Symbol>` to replace functions which have 1-to-1 mappings.
I.e., something like this:
```rust
impl Deref for Symbol {
    type Target = object::Symbol;

    fn deref(&self) -> &Self::Target {
        &self.symbol
    }
}
```